In [ ]:
!pip install transformers torch

In [ ]:
from transformers import pipeline
# 1. Tải pipeline "fill-mask"
# Pipeline này sẽ tự động tải một mô hình mặc định phù hợp (thường là một biến thể của BERT)
mask_filler = pipeline("fill-mask")
# 2. Câu đầu vào với token [MASK]
input_sentence = "Hanoi is the <mask> of Vietnam."
# 3. Thực hiện dự đoán
# top_k=5 yêu cầu mô hình trả về 5 dự đoán hàng đầu
predictions = mask_filler(input_sentence, top_k=5)
# 4. In kết quả
print(f"Câu gốc: {input_sentence}")
for pred in predictions:
  print(f"Dự đoán: '{pred['token_str']}' với độ tin cậy: {pred['score']:.4f}")
  print(f" -> Câu hoàn chỉnh: {pred['sequence']}")

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


Câu gốc: Hanoi is the <mask> of Vietnam.
Dự đoán: ' capital' với độ tin cậy: 0.9341
 -> Câu hoàn chỉnh: Hanoi is the capital of Vietnam.
Dự đoán: ' Republic' với độ tin cậy: 0.0300
 -> Câu hoàn chỉnh: Hanoi is the Republic of Vietnam.
Dự đoán: ' Capital' với độ tin cậy: 0.0105
 -> Câu hoàn chỉnh: Hanoi is the Capital of Vietnam.
Dự đoán: ' birthplace' với độ tin cậy: 0.0054
 -> Câu hoàn chỉnh: Hanoi is the birthplace of Vietnam.
Dự đoán: ' heart' với độ tin cậy: 0.0014
 -> Câu hoàn chỉnh: Hanoi is the heart of Vietnam.


1.  **Mô hình đã dự đoán đúng từ `capital` không?**
    Dựa trên kết quả thực thi trước đó, mô hình đã dự đoán chính xác từ `capital` là lựa chọn hàng đầu để điền vào chỗ trống `<mask>` trong câu "Hanoi is the <mask> of Vietnam." với độ tin cậy rất cao (0.9341).

    ```
    Dự đoán: ' capital' với độ tin cậy: 0.9341
     -> Câu hoàn chỉnh: Hanoi is the capital of Vietnam.
    ```

2.  **Tại sao các mô hình Encoder-only như BERT lại phù hợp cho tác vụ này?**
    Các mô hình Encoder-only như BERT rất phù hợp cho tác vụ *fill-mask* (hay còn gọi là Masked Language Modeling - MLM) vì những lý do sau:

    *   **Hiểu biết ngữ cảnh hai chiều (Bidirectional Context):** Kiến trúc encoder của BERT được thiết kế để xử lý toàn bộ câu đầu vào cùng một lúc, cho phép nó nhìn thấy cả ngữ cảnh bên trái và bên phải của một từ bị che (mask). Điều này rất quan trọng để đoán chính xác từ bị thiếu, vì ý nghĩa của một từ thường phụ thuộc vào các từ xung quanh nó. Ví dụ, trong câu "Hanoi is the <mask> of Vietnam.", BERT có thể sử dụng cả "Hanoi" (bên trái) và "of Vietnam" (bên phải) để suy luận rằng "capital" là từ phù hợp nhất.
    *   **Mô hình tiền huấn luyện trên nhiệm vụ MLM:** BERT được tiền huấn luyện đặc biệt trên nhiệm vụ Masked Language Modeling. Điều này có nghĩa là trong quá trình huấn luyện, một số từ trong câu được che đi (masked), và mô hình được yêu cầu dự đoán các từ đó. Quá trình này giúp BERT học được cách hiểu sâu sắc về mối quan hệ giữa các từ và ngữ pháp trong ngôn ngữ, từ đó rất hiệu quả khi điền vào các chỗ trống.
    *   **Biểu diễn ngữ cảnh (Contextual Embeddings):** BERT tạo ra các vector biểu diễn (embeddings) cho mỗi từ dựa trên ngữ cảnh của nó trong câu. Điều này khác với các phương pháp nhúng từ truyền thống chỉ cung cấp một vector cố định cho mỗi từ. Nhờ đó, BERT có thể phân biệt ý nghĩa của cùng một từ khi nó xuất hiện trong các ngữ cảnh khác nhau, giúp đưa ra dự đoán chính xác hơn cho các từ bị che.

In [ ]:
from transformers import pipeline
# 1. Tải pipeline "text-generation"
# Pipeline này sẽ tự động tải một mô hình phù hợp (thường là GPT-2)
generator = pipeline("text-generation")
# 2. Đoạn văn bản mồi
prompt = "The best thing about learning NLP is"
# 3. Sinh văn bản
# max_length: tổng độ dài của câu mồi và phần được sinh ra
# num_return_sequences: số lượng chuỗi kết quả muốn nhận
generated_texts = generator(prompt, max_length=50, num_return_sequences=1)
# 4. In kết quả
print(f"Câu mồi: '{prompt}'")
for text in generated_texts:
  print("Văn bản được sinh ra:")
  print(text['generated_text'])

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Câu mồi: 'The best thing about learning NLP is'
Văn bản được sinh ra:
The best thing about learning NLP is that we don't have to learn to read it. We can learn by taking a breath. You can just sit there, and read, and the brain will start to understand.

We can learn a lot by listening to music, which is not so much about a desire to read, but we can learn a lot by listening to something from the past. The more we listen to music, the more we will learn to read.

It also works well when we are listening to a song, because we are not used to reading the voice of the choir. We can learn to read songs, but the lyrics, the music, the lyrics are not used for the purpose of reading.

This is great, because when the brain learns a song, it will take a long time to learn the lyrics, so it is much easier to learn a song from the past.

This is not just a matter of studying music. It is also a matter of reading.

It is important to understand that you cannot just listen to music as a means to a 

1.  **Kết quả sinh ra có hợp lý không?**
    Kết quả văn bản được sinh ra từ câu mồi "The best thing about learning NLP is" khá hợp lý. Mô hình GPT-2 đã tiếp tục câu một cách mạch lạc, tạo ra một đoạn văn bản có ý nghĩa, mặc dù đôi khi có thể hơi lạc đề hoặc lặp ý. Cụ thể, mô hình đã mở rộng chủ đề về học NLP và những lợi ích của nó, sau đó chuyển sang các ý tưởng liên quan như học qua nghe nhạc hoặc đọc sách, duy trì một giọng điệu nhất quán. Tuy nhiên, do giới hạn `max_length` được đặt quá thấp (50 token) nhưng sau đó lại bị `max_new_tokens` (256 token) trong quá trình thực thi ghi đè, đoạn văn bản sinh ra khá dài và có thể lặp lại một số ý tưởng.

2.  **Tại sao các mô hình Decoder-only như GPT lại phù hợp cho tác vụ này?**
    Các mô hình Decoder-only như GPT (Generative Pre-trained Transformer) rất phù hợp cho tác vụ sinh văn bản (text generation) vì những lý do sau:

    *   **Kiến trúc tự hồi quy (Autoregressive Architecture):** Kiến trúc của các mô hình Decoder-only được thiết kế để dự đoán token tiếp theo dựa trên tất cả các token đã được sinh ra trước đó trong chuỗi. Điều này tạo ra một quá trình sinh văn bản tự nhiên, từng bước, rất giống với cách con người viết hoặc nói.
    *   **Tiền huấn luyện trên nhiệm vụ Mô hình Ngôn ngữ (Language Modeling):** GPT được tiền huấn luyện trên một lượng lớn dữ liệu văn bản với mục tiêu dự đoán từ tiếp theo trong một chuỗi. Quá trình này giúp mô hình học được ngữ pháp, cú pháp, phong cách và thông tin ngữ nghĩa rộng lớn của ngôn ngữ, cho phép nó tạo ra văn bản mạch lạc và tự nhiên.
    *   **Khả năng tạo văn bản dài và mạch lạc:** Nhờ kiến trúc tự hồi quy và khả năng hiểu ngữ cảnh rộng, các mô hình Decoder-only có thể tạo ra các đoạn văn bản dài, duy trì được sự mạch lạc và tính nhất quán về chủ đề trong suốt quá trình sinh.
    *   **Không có token `[MASK]`:** Khác với các mô hình Encoder-only cần token `[MASK]` để dự đoán từ bị thiếu, mô hình Decoder-only chỉ đơn giản là tiếp tục chuỗi văn bản đã cho, khiến chúng trở nên lý tưởng cho các tác vụ sinh văn bản mở đầu như viết truyện, bài luận, email, hoặc đoạn hội thoại.

In [ ]:
import torch
from transformers import AutoTokenizer, AutoModel

# 1. Chọn một mô hình BERT
model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

# 2. Câu đầu vào
sentences = ["This is a sample sentence."]

# 3. Tokenize câu
# padding=True: đệm các câu ngắn hơn để có cùng độ dài
# truncation=True: cắt các câu dài hơn
# return_tensors='pt': trả về kết quả dưới dạng PyTorch tensors
inputs = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

# 4. Đưa qua mô hình để lấy hidden states
# torch.no_grad() để không tính toán gradient, tiết kiệm bộ nhớ
with torch.no_grad():
  outputs = model(**inputs)

# outputs.last_hidden_state chứa vector đầu ra của tất cả các token
last_hidden_state = outputs.last_hidden_state
# shape: (batch_size, sequence_length, hidden_size)

# 5. Thực hiện Mean Pooling
# Để tính trung bình chính xác, chúng ta cần bỏ qua các token đệm (padding tokens)
attention_mask = inputs['attention_mask']
mask_expanded = attention_mask.unsqueeze(-1).expand(last_hidden_state.size()).float()
sum_embeddings = torch.sum(last_hidden_state * mask_expanded, 1)
sum_mask = torch.clamp(mask_expanded.sum(1), min=1e-9)
sentence_embedding = sum_embeddings / sum_mask

# 6. In kết quả
print("Vector biểu diễn của câu:")
print(sentence_embedding)
print("\nKích thước của vector:", sentence_embedding.shape)

Vector biểu diễn của câu:
tensor([[-6.3874e-02, -4.2837e-01, -6.6779e-02, -3.8430e-01, -6.5784e-02,
         -2.1826e-01,  4.7636e-01,  4.8659e-01,  4.0647e-05, -7.4273e-02,
         -7.4740e-02, -4.7635e-01, -1.9773e-01,  2.4824e-01, -1.2162e-01,
          1.6678e-01,  2.1045e-01, -1.4576e-01,  1.2636e-01,  1.8635e-02,
          2.4640e-01,  5.7090e-01, -4.7014e-01,  1.3782e-01,  7.3650e-01,
         -3.3808e-01, -5.0331e-02, -1.6452e-01, -4.3517e-01, -1.2900e-01,
          1.6516e-01,  3.4004e-01, -1.4930e-01,  2.2422e-02, -1.0488e-01,
         -5.1916e-01,  3.2964e-01, -2.2162e-01, -3.4206e-01,  1.1993e-01,
         -7.0148e-01, -2.3126e-01,  1.1224e-01,  1.2550e-01, -2.5191e-01,
         -4.6374e-01, -2.7261e-02, -2.8415e-01, -9.9249e-02, -3.7017e-02,
         -8.9192e-01,  2.5005e-01,  1.5816e-01,  2.2701e-01, -2.8497e-01,
          4.5300e-01,  5.0945e-03, -7.9441e-01, -3.1008e-01, -1.7403e-01,
          4.3029e-01,  1.6816e-01,  1.0590e-01, -4.8987e-01,  3.1856e-01,
          3.

1.  **Kích thước (chiều) của vector biểu diễn là bao nhiêu? Con số này tương ứng với tham số nào của mô hình BERT?**
    Dựa trên kết quả thực thi trước đó, kích thước của vector biểu diễn (vector nhúng câu) là `768` chiều (`torch.Size([1, 768])`).
    Con số `768` này tương ứng với tham số `hidden_size` (kích thước của không gian ẩn) của mô hình BERT đã được sử dụng (`bert-base-uncased`). Đây là kích thước của các vector biểu diễn đầu ra cho mỗi token từ lớp cuối cùng của mô hình BERT.

2.  **Tại sao chúng ta cần sử dụng `attention_mask` khi thực hiện Mean Pooling?**
    Chúng ta cần sử dụng `attention_mask` khi thực hiện Mean Pooling để đảm bảo rằng các token đệm (padding tokens) không ảnh hưởng đến việc tính toán vector trung bình của câu. Khi xử lý một lô (batch) các câu có độ dài khác nhau, các câu ngắn hơn sẽ được thêm các token đệm để tất cả các câu có cùng độ dài (để có thể xử lý song song bởi mô hình). Các token đệm này không mang thông tin ngữ nghĩa thực sự của câu.

    *   Nếu chúng ta tính trung bình tất cả các token (bao gồm cả token đệm) mà không dùng `attention_mask`, các giá trị của token đệm (thường là vector 0) sẽ làm sai lệch giá trị trung bình, khiến vector biểu diễn cuối cùng không phản ánh chính xác ngữ nghĩa của câu gốc.
    *   `attention_mask` là một tensor nhị phân (chứa 0 hoặc 1), trong đó 1 đánh dấu các token thực sự của câu và 0 đánh dấu các token đệm. Bằng cách nhân `last_hidden_state` (chứa các vector nhúng của tất cả các token) với `attention_mask` (sau khi mở rộng chiều), chúng ta sẽ làm cho các vector nhúng của token đệm trở thành 0. Sau đó, khi tính tổng và chia trung bình, chúng ta chỉ tính toán dựa trên các token thực sự của câu, loại bỏ ảnh hưởng của token đệm.